### PR TITLE
Improve backend implementation of the task module

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
$Subject

Fixes [#59](https://github.com/ballerina-platform/ballerina-standard-library/issues/59)

### CODE REVIEW

File: task-ballerina/src/task/listener.bal [hunk 1]
```diff
@@ -157,22 +162,33 @@ isolated function attachExternal(Listener task, service s, any... attachments) r
...
 isolated function validateConfiguration(TimerConfiguration|AppointmentConfiguration configuration) {
     var noOfRecurrences = configuration[NO_OF_RECURRENCE];
+    var delay = configuration[INITIAL_DELAY];
     var misfirePolicy = configuration.misfirePolicy;
     if (configuration is TimerConfiguration) {
         if (noOfRecurrences is int) {
+            if (noOfRecurrences < 1) {
+                panic ListenerError("Task noOfOccurrences should be a positive integer.");
+            }
             if (noOfRecurrences == 1) {
```
Comment (BuddhiWathsala) [inline task-ballerina/src/task/listener.bal:172]: Better to convert this to a `else if` clause. That will improve performance.

**Review by BuddhiWathsala — CHANGES_REQUESTED** [review]

File: task-native/src/main/java/org/ballerinalang/stdlib/task/actions/TaskActions.java [hunk 2]
```diff
@@ -42,94 +38,93 @@
...
-        Task task = (Task) taskListener.getNativeData(NATIVE_DATA_TASK_OBJECT);
+        String triggerID = (String) taskListener.getNativeData(TaskConstants.TRIGGER_NAME);
+        TaskScheduler taskScheduler = (TaskScheduler) taskListener.getNativeData(TaskConstants.SCHEDULER);
         try {
-            task.resume();
-        } catch (SchedulingException e) {
-            return createTaskError(TaskConstants.SCHEDULER_ERROR, e.getMessage());
+            taskScheduler.resume(triggerID);
+        } catch (SchedulerException e) {
+            return Utils.createTaskError("Cannot resume Listener/Scheduler." + e.getMessage());
```
Comment (ThisaruGuruge) [inline task-native/src/main/java/org/ballerinalang/stdlib/task/actions/TaskActions.java:57]:
```suggestion
            return Utils.createTaskError("Cannot resume the Listener/Scheduler. " + e.getMessage());
```

Check for the other places as well.

File: task-native/src/main/java/org/ballerinalang/stdlib/task/actions/TaskActions.java [hunk 3]
```diff
@@ -42,94 +38,93 @@
...
-            validateService(serviceInformation);
+            Utils.validateService(serviceInformation);
+            BMap<BString, Object> configurations = taskListener.getMapValue(TaskConstants.
+                    MEMBER_LISTENER_CONFIGURATION);
+            taskScheduler.addService(serviceInformation, configurations, triggerID);
         } catch (SchedulingException e) {
-            return createTaskError(e.getMessage());
+            return Utils.createTaskError(e.getMessage());
+        } catch (SchedulerException e) {
+            return Utils.createTaskError("Cannot attach the service to the Listener/Scheduler." + e.getMessage());
```
Comment (ThisaruGuruge) [inline task-native/src/main/java/org/ballerinalang/stdlib/task/actions/TaskActions.java:110]:
```suggestion
            return Utils.createTaskError("Cannot attach the service to the Listener/Scheduler. " + e.getMessage());
```

These error messages seem inconsistent. Check for the other places as well.

File: task-native/src/main/java/org/ballerinalang/stdlib/task/objects/TaskScheduler.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        String name = serviceInformation.getServiceName();
+        JobDataMap jobDataMap = new JobDataMap();
+        jobDataMap.put(TaskConstants.SERVICE_INFORMATION, serviceInformation);
+        JobDetail job = JobBuilder.newJob(TaskJob.class).withIdentity(name).usingJobData(jobDataMap).build();
+        Trigger trigger = Utils.getTrigger(configurations, triggerId);
+        this.jobInfoMap.put(name, job.getKey());
+        this.scheduler.scheduleJob(job, trigger);
+    }
+
+    @SuppressWarnings("unchecked")
```
Comment (ThisaruGuruge) [inline task-native/src/main/java/org/ballerinalang/stdlib/task/objects/TaskScheduler.java:82]: Why do we need this `SuppressWarning` here? If unnecessary, shall we remove it? Check for the other places as well.

File: task-native/src/main/java/org/ballerinalang/stdlib/task/actions/TaskActions.java [hunk 5]
```diff
@@ -42,94 +38,93 @@
...
         BMap<BString, Object> configurations = taskListener.getMapValue(TaskConstants.MEMBER_LISTENER_CONFIGURATION);
-        String configurationTypeName = configurations.getType().getName();
-        Task task;
         try {
-            if (TaskConstants.RECORD_TIMER_CONFIGURATION.equals(configurationTypeName)) {
-                task = processTimer(configurations);
-            } else { // Record type validates at the compile time. Hence, exhaustive validation is not needed.
-                task = processAppointment(configurations);
+            if (!TaskConstants.RECORD_TIMER_CONFIGURATION.equals(configurations.getType().getName())) {
+                Object cronExpression = configurations.get(TaskConstants.MEMBER_CRON_EXPRESSION);
```
Comment (ThisaruGuruge) [inline task-native/src/main/java/org/ballerinalang/stdlib/task/actions/TaskActions.java:120]: Shouldn't this be:

```suggestion
                BString cronExpression = configurations.getStringvalue(TaskConstants.MEMBER_CRON_EXPRESSION);
```

File: task-native/src/main/java/org/ballerinalang/stdlib/task/utils/Utils.java [hunk 6]
```diff
@@ -51,7 +62,7 @@ public static BError createTaskError(String reason, String message) {
     }
 
     @SuppressWarnings("unchecked")
-    public static String getCronExpressionFromAppointmentRecord(Object record) throws SchedulingException {
+    public static String validateCronExpression(Object record) throws SchedulingException {
         String cronExpression = record.toString();
```
Comment (ThisaruGuruge) [inline task-native/src/main/java/org/ballerinalang/stdlib/task/utils/Utils.java:66]:
```suggestion
    public static String validateCronExpression(BString expression) throws SchedulingException {
        String cronExpression = expression.getValue();
```

File: task-native/src/main/java/org/ballerinalang/stdlib/task/utils/Utils.java [hunk 7]
```diff
@@ -87,36 +98,137 @@ private static void validateOnTriggerResource(Type returnParameterType) throws S
...
+                cronScheduleBuilder.withMisfireHandlingInstructionFireAndProceed();
+                break;
+            default:
+        }
+        return cronScheduleBuilder;
+    }
+
+    public static long getMaxRuns(BMap<BString, Object> configurations) {
+        long maxRuns;
+        if (configurations.getIntValue(TaskConstants.FIELD_NO_OF_RUNS) == null) {
```
Comment (ThisaruGuruge) [inline task-native/src/main/java/org/ballerinalang/stdlib/task/utils/Utils.java:209]: This can't be null

**Comment by ThisaruGuruge** [comment]
Just another question, did you check how these changes affect the other stdlib modules, such as Cache?

File: task-ballerina/src/task/commons.bal [hunk 8]
```diff
@@ -94,8 +91,7 @@ public type TimerConfiguration record {|
 #                                        to be fired now by the scheduler.
 public type AppointmentConfiguration record {|
     string cronExpression;
-    int noOfRecurrences?;
-    int thresholdInMillis = 5000;
+    int noOfRecurrences = 0;
```
[the remark this answers is not in the document]
Comment (kalaiyarasiganeshalingam) [reply task-ballerina/src/task/commons.bal:94]: The default value is 0. It executes the job forever. I have added this in the [description](https://github.com/ballerina-platform/module-ballerina-task/pull/31/files/ff437c00c81d35975d3d2d37b91114ca2c7923c2#diff-50766c35f035a8c0ec53a3724fbe150acda7722730f46c6aa4b73e6aa9c2fb16R22).

**Comment by kalaiyarasiganeshalingam** [comment]
> Just another question, did you check how these changes affect the other stdlib modules, such as Cache?

Yes, I have checked the Cache, it has worked.

File: task-ballerina/src/task/commons.bal [hunk 9]
```diff
@@ -19,9 +19,8 @@
 #
 # + intervalInMillis - Timer interval (in milliseconds), which triggers the `onTrigger` resource
 # + initialDelayInMillis - Delay (in milliseconds) after which the timer will run
-# + noOfRecurrences - Number of times to trigger the task after which the task stops running
-# + thresholdInMillis - The number of milliseconds the scheduler will tolerate a trigger to pass its
-#                       next-fire-time by being considered `misfired` before.
+# + noOfRecurrences - Number of times to trigger the task after which the task stops running. If It will be zero,
```
Comment (ThisaruGuruge) [inline task-ballerina/src/task/commons.bal:22]:
```suggestion
# + noOfRecurrences - Number of times to trigger the task after which the task stops running. If It is zero,
```

File: task-native/src/main/java/org/ballerinalang/stdlib/task/utils/TaskConstants.java [hunk 10]
```diff
@@ -68,20 +70,17 @@
...
 
     // Quarts property names
     public static final String QUARTZ_THREAD_COUNT = "org.quartz.threadPool.threadCount";
     public static final String QUARTZ_MISFIRE_THRESHOLD = "org.quartz.jobStore.misfireThreshold";
 
     // Quartz property values.
     public static final String QUARTZ_THREAD_COUNT_VALUE = "10";
+    public static final String QUARTZ_THRESHOLD_VALUE = "5000";
+
+    TaskConstants() {
```
Comment (ThisaruGuruge) [inline task-native/src/main/java/org/ballerinalang/stdlib/task/utils/TaskConstants.java:83]: shouldn't this be `private`?

File: task-ballerina/src/task/commons.bal [hunk 11]
```diff
@@ -78,9 +76,8 @@ public type TimerConfiguration record {|
 #                       Month
 #                       Day-of-Week
 #                       Year (optional field)
-# + noOfRecurrences - Number of times to trigger the task after which the task stops running
-# + thresholdInMillis - The number of milliseconds the scheduler will tolerate a trigger to pass its
-#                       next-fire-time by being considered `misfired` before.
+# + noOfRecurrences - Number of times to trigger the task after which the task stops running. If It will be zero,
```
Comment (MadhukaHarith92) [inline task-ballerina/src/task/commons.bal:79]:
```suggestion
# + noOfRecurrences - Number of times to trigger the task after which the task stops running. If It is zero,
```

File: task-ballerina/src/task/listener.bal [hunk 12]
```diff
@@ -156,23 +156,30 @@ isolated function attachExternal(Listener task, service s, any... attachments) r
...
-
-            } else {
-                if (!(misfirePolicy is RecurringTaskPolicy)) {
-                    panic ListenerError("Wrong misfire policy has given for the repeating execution timer tasks.");
-                }
+        if (noOfRecurrences < 0) {
+            panic ListenerError("Task noOfOccurrences should be a positive integer.");
+        } else if (noOfRecurrences == 1) {
+            if (!(misfirePolicy is OneTimeTaskPolicy)) {
+                panic ListenerError("Wrong misfire policy has given for the one-time execution timer tasks.");
```
Comment (MadhukaHarith92) [inline task-ballerina/src/task/listener.bal:167]:
```suggestion
                panic ListenerError("Wrong misfire policy has been given for the one-time execution timer tasks.");
```

File: task-ballerina/src/task/listener.bal [hunk 13]
```diff
@@ -156,23 +156,30 @@ isolated function attachExternal(Listener task, service s, any... attachments) r
...
             }
+        } else {
+            if (!(misfirePolicy is RecurringTaskPolicy)) {
+                panic ListenerError("Wrong misfire policy has given for the repeating execution timer tasks.");
```
Comment (MadhukaHarith92) [inline task-ballerina/src/task/listener.bal:171]:
```suggestion
                panic ListenerError("Wrong misfire policy has been given for the repeating execution timer tasks.");
```

**Comment by kalaiyarasiganeshalingam** [comment]
The transaction also worked with this changes.

File: task-native/src/main/java/org/ballerinalang/stdlib/task/actions/TaskActions.java [hunk 14]
```diff
@@ -42,94 +38,92 @@
...
-        } catch (SchedulingException e) {
-            return createTaskError(e.getMessage());
+            taskScheduler.stop(triggerID);
+        } catch (SchedulerException e) {
+            return Utils.createTaskError(e.getMessage());
         }
         return null;
     }
 
+    @SuppressWarnings("unchecked")
```
Comment (MadhukaHarith92) [inline task-native/src/main/java/org/ballerinalang/stdlib/task/actions/TaskActions.java:93]: Shall we put an explanatory comment as to why this unchecked warning is being suppressed?

File: task-native/src/main/java/org/ballerinalang/stdlib/task/objects/TaskScheduler.java [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+        jobInfoMap = new HashMap<>();
+    }
+
+    public void stop(String triggerId) throws SchedulerException {
+        Set<TriggerKey> keys = this.scheduler.getTriggerKeys(GroupMatcher.groupEquals(triggerId));
+        if (!keys.isEmpty()) {
+            for (TriggerKey key :keys) {
+                this.scheduler.unscheduleJob(key);
+            }
+
```
Comment (MadhukaHarith92) [inline task-native/src/main/java/org/ballerinalang/stdlib/task/objects/TaskScheduler.java:62]:
```suggestion
```

File: task-native/src/main/java/org/ballerinalang/stdlib/task/utils/Utils.java [hunk 16]
```diff
@@ -87,36 +98,127 @@ private static void validateOnTriggerResource(Type returnParameterType) throws S
...
-        long thresholdInMillis = configurations.getIntValue(TaskConstants.THRESHOLD_IN_MILLIS).intValue();
-        String misfirePolicy = String.valueOf(configurations.getStringValue(TaskConstants.MISFIRE_POLICY));
+    public static SimpleScheduleBuilder createSchedulerBuilder(BMap<BString, Object> configurations) {
+        long interval = configurations.getIntValue(TaskConstants.FIELD_INTERVAL).intValue();
+        long maxRuns = configurations.getIntValue(TaskConstants.FIELD_NO_OF_RUNS).intValue();;
+        String policy = String.valueOf(configurations.getStringValue(TaskConstants.MISFIRE_POLICY));
+        SimpleScheduleBuilder simpleScheduleBuilder = SimpleScheduleBuilder.simpleSchedule()
+                .withIntervalInMilliseconds(interval);
+        if (maxRuns > 0) {
+            // Quartz uses number of repeats but total number of runs are counted here.
```
Comment (MadhukaHarith92) [inline task-native/src/main/java/org/ballerinalang/stdlib/task/utils/Utils.java:132]:
```suggestion
            // Quartz uses the number of repeats but the total number of runs are counted here.
```

File: task-native/src/main/java/org/ballerinalang/stdlib/task/utils/Utils.java [hunk 17]
```diff
@@ -87,36 +98,127 @@ private static void validateOnTriggerResource(Type returnParameterType) throws S
...
+            // Hence, 1 is subtracted from the `maxRuns` to get the repeat count.
```
Comment (MadhukaHarith92) [inline task-native/src/main/java/org/ballerinalang/stdlib/task/utils/Utils.java:133]:
```suggestion
            // Hence, `maxRuns` is subtracted by 1 to get the repeat count.
```

**Review by ThisaruGuruge — APPROVED** [review]
LGTM

**Review by MadhukaHarith92 — APPROVED** [review]

**Review by BuddhiWathsala — APPROVED** [review]

**Pull request merged**